### DISCUSSION TITLE
Top 10 Last Updated

### DISCUSSION THREAD

**Discussion by cweakland** [mcguirepr89/BirdNET-Pi#1098] [Q&A]
On the main page, can one increase the "Top 10 Last Updated" to a value of 50? 

![image](https://github.com/mcguirepr89/BirdNET-Pi/assets/25326695/d8237c6f-1dd7-450a-b68a-6ca4e4e56d8f)

**Comment by stevedee** [comment]
Yes, if you are happy to tinker with the code.

In _scripts/daily_plot.py_ change:-
`# Set number of species to report

readings = 10`
...to required number.

I've made a few display mods...
![BirdNetPiDisplayMods](https://github.com/mcguirepr89/BirdNET-Pi/assets/1870807/fc3fa917-0bc3-4098-8c88-1e2239fbe0bf)

**Reply by cweakland** [reply]
Thank you for the response, it looks like this is just part of the puzzle, any idea on how to disable the line wrap for those longer bird names? 

![image](https://github.com/mcguirepr89/BirdNET-Pi/assets/25326695/353e4869-2484-4d19-8536-b05d19ee2cdf)

**Reply by stevedee** [reply]
The problem is that the text/graph box is a fixed vertical height.

I keep meaning to go back and fix this properly (_so the height grows as the number of detected bird species increases_) but at the time I just bodged it. If you want to join me in this hack, find the section in **_daily_plot.py_** that reads:-

`# Set up plot axes and titles

f, axs = plt.subplots(1, 2, figsize=(10, 4), gridspec_kw=dict(width_ratios=[3, 6]), facecolor='#77C487')
`
...and increase the 2nd digit in the 10, 4 pair to (say) 10, 8 or 10, 12

Just experiment ...or work out how to make it dynamic (_and let me know!_)

**Reply by cweakland** [reply]
Perfect! Thank you! 

![image](https://github.com/mcguirepr89/BirdNET-Pi/assets/25326695/fa042734-0abb-409e-89dd-799e93eac1d5)

**Reply by pddpauw** [reply]
I did the same...
Made changes in daily_plot.py (line 46):
```
# Set number of species to report
readings = 50
```
and also on line 79, where I reduced font type slightly and had more characters on one line (to avoid the species descriptions gets wrapped over 2 lines):
`plot.set_yticklabels(['\n'.join(textwrap.wrap(ticklabel.get_text(), 30)) for ticklabel in plot.get_yticklabels()], fontsize=8)`

Good luck

**Comment by cweakland** [comment]
Thank you both. I combined your suggestions and found this to be pleasing enough, I included a diff patch to help others. Also, I amended the "Bottom" chart as well. Now to hunt that Ceder Waxwing!! I want to see him/her! 
 
![image](https://github.com/mcguirepr89/BirdNET-Pi/assets/25326695/26c88ef8-44b3-40a0-9c0e-f8925ec1e8d8)

[patch.txt](https://github.com/mcguirepr89/BirdNET-Pi/files/13850526/patch.txt)

**Reply by stevedee** [reply]
I was feeling a bit guilty about the hack, so here is a quick attempt to make the grid appear proportional.

In _daily_plot.py_ find the number of unique species for today (I've been using this for a while):-

**#===============
cursor.execute('SELECT DISTINCT (Com_Name) FROM detections WHERE Date = DATE(\'now\', \'localtime\')')
uTable_rows = cursor.fetchall()
#===============**

Find number of display rows:-

**# Set number of top species to report
readings = 25
#===============
if uTable_rows < readings:
	gridMultiplier=uTable_rows
else:
	gridMultiplier=readings
#===============**	

Adjust row/grid height:-

**# Set up plot axes and titles
#===============
f, axs = plt.subplots(1, 2, figsize=(10, (0.3 * gridMultiplier)), gridspec_kw=dict(width_ratios=[3, 6]), facecolor='#77C487')
#===============**

**Reply by pddpauw** [reply]
nice one!

**Reply by cweakland** [reply]
I did not have any luck with this, could you share your whole daily_plot.py?

**Reply by stevedee** [reply]
I'm sorry, I've been trying to renovate my house and write code at the same time, _and making a bad job of both!_

Code between #==== is the new or modified code.

In _daily_plot.py_ find the number of unique species for today:-

conn = sqlite3.connect(userDir + '/BirdNET-Pi/scripts/birds.db')
df = pd.read_sql_query("SELECT * from detections", conn)
cursor = conn.cursor()

#===============
cursor.execute('SELECT DISTINCT (Com_Name) FROM detections WHERE Date = DATE(\'now\', \'localtime\')')
uTable_rows = len(cursor.fetchall())
#===============

Find number of display rows:-
`# Set number of top species to report`
readings = 25
#===============
if uTable_rows > readings:
    gy=readings
else:
    gy=uTable_rows
     
gy = 0.4 * gy
#===============

Adjust row/grid height:-

`# Set up plot axes and titles`
#===============
f, axs = plt.subplots(1, 2, figsize=(10, gy), gridspec_kw=dict(width_ratios=[3, 6]), facecolor='#77C487')
#===============


**_Important:_** in Python, indentation is important, but I can't get this message to format properly using text or the code delimiters "<>"
So please indent the 2 lines "gy=readings" and "gy=uTable_rows"

**Reply by cweakland** [reply]
Still no luck here, take a look at this diff, did I do it right?
[patch.txt](https://github.com/mcguirepr89/BirdNET-Pi/files/13855093/patch.txt)

**Reply by stevedee** [reply]
A couple of things, take a second look at my last post;

1. You are missing: len()
uTable_rows = len(cursor.fetchall())

2. the "if" comparator should be ">"
if uTable_rows > readings:

I hope that helps.

**Comment by cweakland** [comment]
Perfect thank you! Here is a diff patch for anyone interested, take a look at line 7, I had to escape the "now" and "localtime".

`cursor.execute('SELECT DISTINCT (Com_Name) FROM detections WHERE Date = DATE(\'now\', \'localtime\')')`

[patch.txt](https://github.com/mcguirepr89/BirdNET-Pi/files/13863186/patch.txt)

![image](https://github.com/mcguirepr89/BirdNET-Pi/assets/25326695/a41f4b69-a304-48f8-a6a9-de61e0c936e8)

**Reply by Fish31520** [reply]
Hi,

Noob´s question here but how should I use your patch.txt file ? Should I copy it and add a specific command somewhere or just paste what it contains in my daily_plot.py file ?

Thanks !

**Reply by cweakland** [reply]
Sure thing, copy the patch.txt into your BirdNET/scripts directory, then do:

cd /home/pi/BirdNET-Pi/scripts

#First make a backup of your origional file:
cp daily_plot.py daily_plot.py_orig

#Then patch it
patch daily_plot.py patch.txt 

It takes a minute for the graphs to be re-drawn, good luck.

**Reply by Fish31520** [reply]
@cweakland Thanks a lot !

I've had a small issue :

`patching file daily_plot.py
Hunk #2 FAILED at 46.
Hunk #3 FAILED at 58.
2 out of 6 hunks FAILED -- saving rejects to file daily_plot.py.rej
`

Any clue ?

**Reply by cweakland** [reply]
Sounds like your daily_plot.py is different. I just did this procedure two times on new installs for friends. Try and download the original daily_plot.py and patch it:

wget https://raw.githubusercontent.com/mcguirepr89/BirdNET-Pi/main/scripts/daily_plot.py -O daily_plot.py
patch daily_plot.py patch.txt

**Reply by Fish31520** [reply]
You were right, the issue came from the daily_plot.py that I tried to modify but now with your help it seems to work fine !
Thanks a lot !

**Comment by ClemensGruber** [comment]
Would be great to have this as a variable in the settings, at the moment you can see by default just the top ten or last ten, so a complete list would be nice!
